compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs

```diff
@@ -29,26 +30,50 @@ import DA.Daml.LF.Ast.Optics (_PRSelfModule, dataConsType)
 import DA.Daml.LF.TypeChecker.Env
 import DA.Daml.LF.TypeChecker.Error
 
+-- This is only used during serializability inference. During typechecking the world
+-- contains the current module.
+data CurrentModule = CurrentModule
+  { modName :: ModuleName
+  , modInterfaces :: HS.HashSet TypeConName
+  }
+
 -- | Determine whether a type is serializable. When a module name is given,
 -- data types in this module are returned rather than lookup up in the world.
 -- If no module name is given, the returned set is always empty.
 serializabilityConditionsType
   :: World
   -> Version
-  -> Maybe ModuleName
+  -> Maybe CurrentModule
      -- ^ See description on `serializabilityConditionsDataType`.
   -> HS.HashSet TypeVarName
      -- ^ Type variables that are bound by a surrounding data type definition
      -- if any. The check that all of them are of kind '*' must be performed by
      -- the caller.
   -> Type
   -> Either UnserializabilityReason (HS.HashSet TypeConName)
-serializabilityConditionsType world0 _version mbModName vars = go
+serializabilityConditionsType world0 _version mbCurrentModule vars = go
   where
     noConditions = Right HS.empty
     go = \case
       -- This is the only way 'ContractId's, 'List's and 'Optional's are allowed. Other cases handled below.
-      TContractId typ -> go typ
+      TContractId typ
+          -- While an interface payload I is not serializable, ContractId I
+          -- is so specialcase this here.
+          | TCon con <- typ -> case (lookupDataType con world0, mbCurrentModule) of
+              (Right t, _) -> case dataCons t of
+                DataInterface -> noConditions
+                _ -> go typ
+              (Left _, Just currentModule)
+                | Right tconName <- matching (_PRSelfModule $ modName currentModule) con
+                -> if tconName `HS.member` modInterfaces currentModule
+                     then noConditions
+                     else go typ
+              (Left err, _) -> error $ showString "Serializability.serializabilityConditionsDataTyp: " $ show err
+          | TCon con <- typ
+          , Just currentModule <- mbCurrentModule
+          , Right tconName <- matching (_PRSelfModule $ modName currentModule) con
+          , tconName `HS.member` modInterfaces currentModule -> noConditions
+          | otherwise -> go typ
```

Comment:
Just a suggestion, I would try to factor out an `isInterface` here that just determines if typ refers to an interface type constructor.

```suggestion
      TContractId typ ->
          if isInterface world0 mbCurrentModule typ
            then noConditions
            else go typ
```

Reply:
good idea done